Handle const and static local variables

333fred opened the pull request:
Adds support for const and static local variables, as decided in the design meeting a few weeks ago. Notes from the meeting:

> **Initialization of constant locals**
> Decision: remove initialization of constant locals from the control flow graph.

> **Initialization of static VB locals**
> Decision: Add conditional branch that jumps over the initialization assignment. An instance of a new node, IStaticLocalInitializationSemaphore should be used as the branch condition and should refer to the local symbol. BasicBlocks involved into initialization should be in a special region, new region kind should be added.

Tagging @dotnet/roslyn-compiler @dotnet/analyzer-ioperation @AlekseyTs for review.

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/TestOperationVisitor.cs:
```diff
@@ -1132,5 +1132,12 @@ public override void VisitCaughtException(ICaughtExceptionOperation operation)
+
+        public override void VisitStaticLocalInitialzationSemaphore(IStaticLocalInitializationSemaphoreOperation operation)
+        {
+            Assert.Equal(OperationKind.StaticLocalInitializationSemaphore, operation.Kind);
```
>Assert.Equal(OperationKind.StaticLocalInitializationSemaphore, operation.Kind); [](start = 12, length = 79)

Let's also assert that the local is not null and is a static local

AlekseyTs commented on src/Test/Utilities/Portable/Compilation/ControlFlowGraphVerifier.cs:
```diff
@@ -299,6 +299,12 @@ void enterRegions(ControlFlowGraph.Region region, int firstBlockOrdinal)
                         break;
+
+                    case ControlFlowGraph.RegionKind.StaticLocalInitializer:
+                        Assert.Null(region.ExceptionType);
```
>Assert.Null(region.ExceptionType); [](start = 24, length = 34)

Also ```Assert.Empty(region.Locals);```?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb:
```diff
@@ -3085,25 +3085,32 @@ Block[B0] - Entry
-                ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1) (Syntax: '1')
+        Statements (0)
+        Jump if True (Regular) to Block[B2]
+            IStaticLocalInitializationSemaphoreOperation (Local Symbol: b As System.Int32) (OperationKind.StaticLocalInitializationSemaphore, Type: null, Constant: null, IsImplicit) (Syntax: 'b')
```
>null [](start = 148, length = 4)

I think the type should be Boolean

AlekseyTs commented on src/Compilers/Core/Portable/Operations/IStaticLocalInitializationSemaphoreOperation.cs:
```diff
@@ -0,0 +1,15 @@
+namespace Microsoft.CodeAnalysis.Operations
+{
+    /// <summary>
+    /// Represents the check during initialization of a VB static local that is initialized on the first call of the function, and never again.
```
>Represents the check during initialization of a VB static local that is initialized on the first call of the function, and never again. [](start = 8, length = 135)

I think we should clarify what true/false values mean for this node.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1998,13 +2015,29 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
+                    {
+                        BasicBlock initializerBlock = CurrentBasicBlock;
+
+                        var initializationSemaphore = new StaticLocalInitializationSemaphoreOperation(localSymbol, localSymbol.DeclaringSyntaxReferences.Single().GetSyntax(), type: null, constantValue: null, isImplicit: true);
```
>null [](start = 181, length = 4)

semanticModel.Compilation.GetSpecialType(SpecialType.System_Boolean)?

AlekseyTs commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb:
```diff
@@ -3468,6 +3475,245 @@ Block[B2] - Exit
+    Block[B1] - Block
+        Predecessors: [B0]
+        Statements (0)
+        Jump if True (Regular) to Block[B5]
```
>B5 [](start = 40, length = 2)

This should be B2

333fred replied:
Fixed.

---
In reply to: [174604311](https://github.com/dotnet/roslyn/pull/25458#discussion_r174604311) [](ancestors = 174604311)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1998,13 +2015,29 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
+                        _currentBasicBlock = null;
+                        LeaveRegion();
+                        LinkBlocks(previousBlock, CurrentBasicBlock);
+                        LinkBlocks(previousBlock, (initializationSemaphore, JumpIfTrue: true, RegularBranch(initializerBlock)));
```
>RegularBranch(initializerBlock) [](start = 94, length = 31)

This doesn't look right. This links to the very last block in the initializer, when we should be linking to the first one.
Also, I suggest to move all code dealing with the Semaphor "business" right before we enter initialization region. Allocate new block (afterInitialization) there and link it with the CurrentBasicBlock as appropriate, enter the region and link the conditional branch. Here, leave the region and add the afterInitialization block.

AlekseyTs commented:
Done with review pass (iteration 3)

333fred commented:
@AlekseyTs addressed feedback.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1969,6 +1970,30 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
+                    LinkBlocks(CurrentBasicBlock, afterInitialization);
+
+                    ITypeSymbol booleanType = ((Operation)operation).SemanticModel.Compilation.GetSpecialType(SpecialType.System_Boolean);
+                    var initializationSemaphore = new StaticLocalInitializationSemaphoreOperation(localSymbol, localSymbol.DeclaringSyntaxReferences.Single().GetSyntax(), booleanType, constantValue: null, isImplicit: true);
```
>localSymbol.DeclaringSyntaxReferences.Single().GetSyntax() [](start = 111, length = 58)

I don't think it is a good idea to make an assumption that we can always get syntax like that, we should simply use syntax from the IOperation nodes at hand.

333fred replied:
Fixed.

---
In reply to: [174859099](https://github.com/dotnet/roslyn/pull/25458#discussion_r174859099) [](ancestors = 174859099)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1969,6 +1970,30 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
+                    Operation.SetParentOperation(initializationSemaphore, null);
+
+                    var initializerBlock = new BasicBlock(BasicBlockKind.Block);
+                    LinkBlocks(CurrentBasicBlock, (initializationSemaphore, JumpIfTrue: true, RegularBranch(initializerBlock)));
```
>CurrentBasicBlock [](start = 31, length = 17)

The code uses CurrentBasicBlock property twice with the intent to work with the same basic block. However, the property itself can allocate and return new blocks. I think the code would be clearer and more robust if the property was used only once and the result was cached in a local.

333fred replied:
With the new structure I'm only calling `CurrentBasicBlock` once, so this isn't necessary.

---
In reply to: [174860738](https://github.com/dotnet/roslyn/pull/25458#discussion_r174860738) [](ancestors = 174860738)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1969,6 +1970,30 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
+                    LinkBlocks(CurrentBasicBlock, (initializationSemaphore, JumpIfTrue: true, RegularBranch(initializerBlock)));
+
+                    EnterRegion(new RegionBuilder(ControlFlowGraph.RegionKind.StaticLocalInitializer));
+                    AppendNewBlock(initializerBlock, linkToPrevious: false);
```
>AppendNewBlock(initializerBlock, linkToPrevious: false); [](start = 20, length = 56)

Why ```linkToPrevious: false```? Also, I think the code would be simpler and more robust if we were not trying to allocate and add the first block in region explicitly, but instead let the machinery to do the allocation, addition and linking. In order to achieve that, we only link conditional branch above (this will also take care of double use of CurrentBasicBlock property), but we change its kind to JumpIfFalse and link it to afterInitialization. Then we simply enter the region and we are done, no need to allocate and add initializerBlock explicitly.

333fred replied:
We don't want to link to previous because there's no default branches to the initializer block. The only way to get to it is via the conditional set just above.

---
In reply to: [174864732](https://github.com/dotnet/roslyn/pull/25458#discussion_r174864732) [](ancestors = 174864732)

333fred replied:
Hmm. It struck me as more natural to have the conditional branch to be to the initializer, as I imagined the code would look more like this:
```
If (!initialized)
    ' Initialize
End If
```

I can give the inverted structure a try though.

---
In reply to: [174932845](https://github.com/dotnet/roslyn/pull/25458#discussion_r174932845) [](ancestors = 174932845,174864732)

333fred replied:
Changed to use this structure.

---
In reply to: [174933838](https://github.com/dotnet/roslyn/pull/25458#discussion_r174933838) [](ancestors = 174933838,174932845,174864732)

AlekseyTs commented:
Done with review pass (iteration 4)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [373459201](https://github.com/dotnet/roslyn/pull/25458#issuecomment-373459201) [](ancestors = 373459201)

333fred commented:
@dotnet/roslyn-compiler @dotnet/analyzer-ioperation can I get another review from each, please

cston commented on src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs:
```diff
@@ -6531,4 +6531,30 @@ public override void Accept(OperationVisitor visitor)
+    internal sealed class StaticLocalInitializationSemaphoreOperation : Operation, IStaticLocalInitializationSemaphoreOperation
+    {
+        public StaticLocalInitializationSemaphoreOperation(ILocalSymbol local, SyntaxNode syntax, ITypeSymbol type, Optional<object> constantValue, bool isImplicit) :
+            base(OperationKind.StaticLocalInitializationSemaphore, semanticModel: null, syntax, type, constantValue, isImplicit)
```
Are `constantValue` and `isImplicit` necessary, or are the values always `null` and `true`?

333fred replied:
They're always the same, I'll remove the parameters.

---
In reply to: [175159578](https://github.com/dotnet/roslyn/pull/25458#discussion_r175159578) [](ancestors = 175159578)

333fred replied:
Good catch, they should have been `default` and true. I always miss that `null` can be given to an `Optional` and not do what I would have expected it to.

---
In reply to: [175165195](https://github.com/dotnet/roslyn/pull/25458#discussion_r175165195) [](ancestors = 175165195,175159578)

cston commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1998,13 +2021,22 @@ private void HandleVariableDeclaration(IVariableDeclarationOperation operation)
                 }
 
+                // If we have an afterInitialization, then we must have an initializer to ensure we don't create empty regions that can't be cleaned up.
+                Debug.Assert(afterInitialization == null || initializer != null);
```
Perhaps add `Debug.Assert((afterInitialization != null) == localSymbol.IsStatic);` since it looks like we're ignoring `afterInitialization` when `localSymbol` is not a static.

333fred replied:
Can't be quite this, as you can have a null afterInitialization if there's no initializer for the static variable. I'll add something similar.

---
In reply to: [175163581](https://github.com/dotnet/roslyn/pull/25458#discussion_r175163581) [](ancestors = 175163581)

cston commented on src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IVariableDeclaration.vb:
```diff
@@ -3468,6 +3475,245 @@ Block[B2] - Exit
+]]>.Value
+
+            VerifyFlowGraphAndDiagnosticsForTest(Of MethodBlockSyntax)(source, expectedFlowGraph, expectedDiagnostics)
+        End Sub
```
Consider testing two `Static` locals in the same scope.

333fred commented:
@cston addressed feedback.
